core/primitives/src/block_body.rs

```diff
@@ -17,61 +19,104 @@ pub struct BlockBodyV1 {
+
+    // Data to confirm the correctness of randomness beacon output
+    pub vrf_value: Value,
+    pub vrf_proof: Proof,
+
+    // Chunk endorsements
+    // These are structured as a vector of Signatures from all ordered chunk_validators
+    // for each shard got from fn get_ordered_chunk_validators
+    // chunk_endorsements[shard_id][chunk_validator_index] is the signature (if present).
+    // If the chunk_validator did not endorse the chunk, the signature is None.
+    pub chunk_endorsements: Vec<Vec<Option<Box<Signature>>>>,
+}
+
 // For now, we only have one version of block body.
 // Eventually with ChunkValidation, we would include ChunkEndorsement in BlockBodyV2
 #[derive(BorshSerialize, BorshDeserialize, Clone, PartialEq, Eq, Debug)]
 pub enum BlockBody {
     V1(BlockBodyV1),
+    V2(BlockBodyV2),
 }
 
 impl BlockBody {
     pub fn new(
-        _protocol_version: ProtocolVersion,
+        protocol_version: ProtocolVersion,
         chunks: Vec<ShardChunkHeader>,
         challenges: Challenges,
         vrf_value: Value,
         vrf_proof: Proof,
+        chunk_endorsements: Vec<Vec<Option<Box<Signature>>>>,
     ) -> Self {
-        // Eventually we will have different versions of block body, but for now we only have V1.
-        BlockBody::V1(BlockBodyV1 { chunks, challenges, vrf_value, vrf_proof })
+        if !checked_feature!("stable", ChunkValidation, protocol_version) {
+            BlockBody::V1(BlockBodyV1 { chunks, challenges, vrf_value, vrf_proof })
+        } else {
+            BlockBody::V2(BlockBodyV2 {
+                chunks,
+                challenges,
+                vrf_value,
+                vrf_proof,
+                chunk_endorsements,
+            })
+        }
     }
 
     #[inline]
     pub fn chunks(&self) -> &[ShardChunkHeader] {
         match self {
             BlockBody::V1(body) => &body.chunks,
+            BlockBody::V2(body) => &body.chunks,
         }
     }
 
     #[inline]
     pub fn challenges(&self) -> &Challenges {
         match self {
             BlockBody::V1(body) => &body.challenges,
+            BlockBody::V2(body) => &body.challenges,
         }
     }
 
     #[inline]
     pub fn vrf_value(&self) -> &Value {
         match self {
             BlockBody::V1(body) => &body.vrf_value,
+            BlockBody::V2(body) => &body.vrf_value,
         }
     }
 
     #[inline]
     pub fn vrf_proof(&self) -> &Proof {
         match self {
             BlockBody::V1(body) => &body.vrf_proof,
+            BlockBody::V2(body) => &body.vrf_proof,
+        }
+    }
+
+    #[inline]
+    pub fn chunk_endorsements(&self) -> &[Vec<Option<Box<Signature>>>] {
+        match self {
+            BlockBody::V1(_) => &[],
+            BlockBody::V2(body) => &body.chunk_endorsements,
         }
     }
 
     pub fn compute_hash(&self) -> CryptoHash {
         match self {
             BlockBody::V1(body) => body.compute_hash(),
+            BlockBody::V2(body) => CryptoHash::hash_borsh(body),
```

Comment:
for versions other than V1, should we hash the entire BlockBody rather than the inner struct? That way the version itself is included into the hash (so somebody can't fake a BlockBody with the wrong version if somehow it aliases the same borsh encoding).

We also would need to make sure that V1 is only ever included in the V3 version, and when validating a block body for the V3 version, any other version will be rejected, and when validating a block body for a > V3 version, V1 would be rejected as well. So in Chain::verify_block_hash_and_signature I think we need to insert some additional checks. It would be nice if we could also test this, though we might only be able to do that to a limited extent (since getting the hash to match might be impossible).

What are your thoughts?

Reply:
So, I ended up adding a check for block.chunk_endorsements() being non-empty for protocol_version > ChunkValidation.

This should let us automatically verify that BlockBodyV1 is not present in BlockV4 as otherwise chunk_endorsements would be empty.

Not sure how to write a test for this tho.